src/nvmet.c

```diff
@@ -833,209 +909,230 @@ int nvmet_revalidate_size(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char
  */
 int nvmet_unexport_volume(char *device, char *host, int port, char *return_message)
 {
-	int rc = INVALID_VALUE, n, err, hostno = 0, allowed_host_number = 0;
+	int rc = INVALID_VALUE, n, allowed_host_number, rv = INVALID_VALUE, port_number;
 	FILE *fp;
-	char hostname[0x40] = {0x0}, path[NAMELEN] = {0x0};
+	char hostname[0x40] = {0x0}, path[NAMELEN] = {0x0}, nqn[NAMELEN] = {0x0};
+	mode_t mode = (S_IRUSR | S_IWUSR | S_IXUSR | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
 	struct dirent **list;
 	char *msg;
-	char json_message[NAMELEN] = {0};
-	char json_message_temp[NAMELEN] = {0};
+	char target_string[NAMELEN] = {0};
+	char target_string_final[NAMELEN] = {0};
+	char regexp[NAMELEN] = {0};
+	struct NVMET_PROFILE *profile;
+	struct NVMET_PROFILE *orig_profile;
+	struct NVMET_PORTS *ports;
+	struct NVMET_PORTS *orig_ports;
 
-	gethostname(hostname, sizeof(hostname));
-	sprintf(path, "%s/%s%s-%s", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
-	if (access(path, F_OK) != SUCCESS) {
-		sprintf(path, "%s%s-%s", NQN_HDR_STR, hostname, device);
-		msg = "Error. NQN export: %s does not exist.";
-		print_error(msg, return_message, path);
+	if (gethostname(hostname, sizeof(hostname)) < 0) {
+		msg = "Error. Unable to obtain hostname. %s: gethostname: %s";
+		print_error(msg, return_message, __func__, strerror(errno));
 		return rc;
 	}
 
-	/* Check if namespace 1 exists. That is the only namespace we define. */
-	sprintf(path, "%s/%s%s-%s/namespaces/1", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
-	if (access(path, F_OK) != SUCCESS) {
-		msg = "%s: A RapidDisk defined namespace does not exist.";
-		print_error(msg, return_message, __func__);
-		return rc;
-	}
+	sprintf(nqn, "%s%s-%s", NQN_HDR_STR, hostname, device);
 
-	/* Make sure that no other namespaces exist. We do not create anything higher than 1. */
-	sprintf(path, "%s/%s%s-%s/namespaces/", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
-	if ((err = scandir(path, &list, scandir_filter_no_dot, NULL)) < 0) {
-		msg = "Error. Unable to access %s. %s: scandir: %s";
-		print_error(msg, return_message, path, __func__, strerror(errno));
-		return INVALID_VALUE;
-	}
+	if (port != INVALID_VALUE) {
+		ports = nvmet_scan_ports(return_message, &rv);
```

Comment:
@matteotenca How about this section? It looks like all of this work can be replace with:
                sprintf(path, "%s/%d", SYS_NVMET_PORTS, port);
                if (access(path, F_OK) != SUCCESS) {
                        sprintf(return_msg, "Error. NVMe Target Port %d does not exists.", port);
                        return INVALID_VALUE;
                }
                sprintf(path, "%s/%d/subsystems/%s", SYS_NVMET_PORTS, port, nqn);
                if (access(path, F_OK) != SUCCESS) {
                        sprintf(return_msg, "Error. NQN %s does not exist on NVMe Target Port %d.", nqn, port);
                        return INVALID_VALUE;
                }

Thoughts?

Reply:
Same as in https://github.com/pkoutoupis/rapiddisk/pull/153#discussion_r1118143204